File: FoodDetectionApp/AI/AIProvider.swift
import Foundation
import UIKit

/// Abstraction over AI inference backends (Gemini Cloud, Gemma On-Device).
/// Both `GeminiCloudProvider` and `GemmaLocalProvider` implement this.
protocol AIProvider {
    var providerName: String { get }

    func analyzeFood(image: UIImage) async throws -> (name: String, info: NutritionInfo)
    func searchFood(query: String) async throws -> (name: String, info: NutritionInfo)
    func getCoachAdvice(userStats: UserStats?, logs: [NutritionManager.FoodLog],
                        healthData: String, historyTOON: String, userQuery: String) async throws -> String
    func generateMealPlan(userStats: UserStats?, calorieBudget: Int) async throws -> [PlannedMeal]
    func estimatePortion(image: UIImage) async throws -> (name: String, info: NutritionInfo, estimatedGrams: Double)
    func compareBeforeAfter(before: UIImage, after: UIImage) async throws -> String
    func ocrNutritionLabel(image: UIImage) async throws -> (name: String, info: NutritionInfo)
    func generateInsights(foodHistory: String, calorieBudget: Int, userStats: UserStats?) async throws -> String
}

File: android/app/src/main/java/com/foodsense/android/services/ai/AIProvider.kt
package com.foodsense.android.services.ai

import android.graphics.Bitmap
import com.foodsense.android.data.FoodLog
import com.foodsense.android.data.MealPlanDay
import com.foodsense.android.data.NutritionInfo
import com.foodsense.android.data.QuizQuestion
import com.foodsense.android.data.UserStats
import com.foodsense.android.data.WeeklyInsight

interface AIProvider {
    val providerName: String

    suspend fun analyzeFood(image: Bitmap): Pair<String, NutritionInfo>
    suspend fun searchFood(query: String): Pair<String, NutritionInfo>
    suspend fun getCoachAdvice(
        userStats: UserStats?, logs: List<FoodLog>,
        healthData: String, historyToon: String, userQuery: String
    ): String
    suspend fun generateMealPlan(userStats: UserStats?, calorieBudget: Int): List<MealPlanDay>
    suspend fun getWeeklyInsights(userStats: UserStats?, history: String, healthData: String): WeeklyInsight
    suspend fun getQuizQuestion(): QuizQuestion
    suspend fun predictWeight(userStats: UserStats?, history: String): String
}
